### ISSUE TITLE
include_role module - Dynamic includes inside roles

### ISSUE THREAD

**Issue by boxrick** [ansible/ansible#21328]
##### ISSUE TYPE

 - Bug Report

##### COMPONENT NAME

include_role module - Dynamic includes inside roles

##### ANSIBLE VERSION

```
ansible 2.3.0.0
  config file =
  configured module search path = Default w/o overrides
```

##### OS / ENVIRONMENT

Ubuntu 16.04
MacOSX 10.xx

##### SUMMARY
When using include_role module and dynamic includes inside the role, rather than using the correct path within the role, it uses the path where the playbook lives. As a result this generates an error. 

##### STEPS TO REPRODUCE

Top level playbook
```yaml
- hosts: hosts
  become: yes
  become_method: sudo
  vars:
    user_list: "{{ users_and_stuff }}"
  pre_tasks:
    - include: general-base.yml
  roles:
  - { role: specific_install_1 }
  - { role: specific_install_2 }
```

general-base.yml
```yaml
---
- name: Include proxy role
  include_role:
      name: proxy_install

- name: Include packages role
  include_role:
      name: base-packages

- name: Include hosts setup role
  include_role:
      name: base-hosts
```
proxy_install/tasks/main.yml
```yaml
---
### Works fine
- include: install.yml

### Broken
- name: do stuff
  include: do_stuff.yml
  with_items:
    - "{{ user_list }}"
```

##### EXPECTED RESULTS

do_stuff.yml to be included and iterated over with user_list

##### ACTUAL RESULTS
```
fatal: [host]: FAILED! => {
    "failed": true,
    "reason": "the file_name 'TOPLEVELPLAYBOOKPATH/dostuff.yml' does not exist, or is not readable"
}
```

**Comment by bcoca** [comment]
where is do_stuff.yml?

**Comment by boxrick** [comment]
Apologies the error was wrong I updated this to be the correct name:

The path dostuff is actually in as as follows: roles/proxy_install/tasks/dostuff.yml

**Comment by boxrick** [comment]
This is still an issue in version 2.3 and renders include_role useless when using loops inside roles.

**Comment by bcoca** [comment]
cc @jimi-c

**Comment by boxrick** [comment]
##### Example of command:

Run from top level 

```ansible-playbook -i inventory ./playbooks/baremetal-deploy/playbook.yml```

#####  Git repo with example:
[https://github.com/boxrick/bugfix](https://github.com/boxrick/bugfix)

##### Error output:

```
PLAY [host] ****************************************************************************************************************************************************************

TASK [Gathering Facts] ************************************************************************************************************************************************************
Friday 19 May 2017  14:36:57 +0000 (0:00:00.135)       0:00:00.135 ************
ok: [host]

TASK [working_role : Hello] *******************************************************************************************************************************************************
Friday 19 May 2017  14:36:58 +0000 (0:00:01.032)       0:00:01.167 ************
ok: [host] => {
    "changed": false,
    "msg": "Hello I work hello"
}

TASK [broken_role : include] ******************************************************************************************************************************************************
Friday 19 May 2017  14:36:58 +0000 (0:00:00.122)       0:00:01.289 ************
fatal: [host]: FAILED! => {"failed": true, "reason": "the file_name '/path/playbooks/baremetal-deploy/broken_task.yml' does not exist, or is not readable"}
fatal: [host]: FAILED! => {"failed": true, "reason": "the file_name '/path/playbooks/baremetal-deploy/broken_task.yml' does not exist, or is not readable"}
	to retry, use: --limit @/path/playbooks/baremetal-deploy/playbook.retry

PLAY RECAP ************************************************************************************************************************************************************************
host                    : ok=2    changed=0    unreachable=0    failed=2

Friday 19 May 2017  14:36:58 +0000 (0:00:00.192)       0:00:01.481 ************
===============================================================================
Gathering Facts --------------------------------------------------------- 1.03s
broken_role : include --------------------------------------------------- 0.19s
working_role : Hello ---------------------------------------------------- 0.12s
```

**Comment by mwinfie** [comment]
@boxrick Does your dynamic include work in the scenario where you aren't using include_role? I'm running into this same error, but I'm using a dynamic include in my role and running a playbook that simply applies that role.

**Comment by boxrick** [comment]
Yes it works perfectly fine when I don't use include role. In fact it is even demonstrated in the above repo with the use of the normal role includes which occurs after the inclusive base role.

**Comment by mwinfie** [comment]
@boxrick your working_role is doing a static include of working_task.yml. I pulled down your example and ran it without any changes, it worked exactly as you described above.

I then modified working_role/tasks/main.yml...
`---
- include: working_task.yml
  static: no`

and reran the test and got this result...
`PLAY [host] ***************************************************************************************


TASK [Gathering Facts] *******************************************************************************

ok: [host]

TASK [working_role : include] ************************************************************************

fatal: [host]: FAILED! => {"failed": true, "reason": "the file_name '/vagrant/projects/bugfix/playbooks/baremetal-deploy/working_task.yml' does not exist, or is not readable"}
        to retry, use: --limit @/vagrant/projects/bugfix/playbooks/baremetal-deploy/playbook.retry

PLAY RECAP *******************************************************************************************

host                       : ok=1    changed=0    unreachable=0    failed=1`

I believe that dynamic includes are failing in this way even if include_role is not being used.

**Comment by mwinfie** [comment]
@boxrick your working_role is doing a static include of working_task.yml. I pulled down your example and ran it without any changes, it worked exactly as you described above.

I then modified working_role/tasks/main.yml...
```
---
- include: working_task.yml
  static: no
```

and reran the test and got this result...
```
PLAY [host] ***************************************************************************************


TASK [Gathering Facts] *******************************************************************************

ok: [host]

TASK [working_role : include] ************************************************************************

fatal: [host]: FAILED! => {"failed": true, "reason": "the file_name '/vagrant/projects/bugfix/playbooks/baremetal-deploy/working_task.yml' does not exist, or is not readable"}
        to retry, use: --limit @/vagrant/projects/bugfix/playbooks/baremetal-deploy/playbook.retry

PLAY RECAP *******************************************************************************************

host                       : ok=1    changed=0    unreachable=0    failed=1
```

I believe that dynamic includes are failing in this way even if include_role is not being used.

**Comment by mwinfie** [comment]
@boxrick also in case you weren't aware, an include is dynamic by default when you use it with_items. You don't have to specifically set "static: no".

**Comment by rcolanero** [comment]
Seeing the same behavior in a new role using Ansible 2.3.0.0. Role has an include tasks/foo.yml in the role's tasks folder. When calling the role via "roles:" at start of a play it works fine but when I use "include_role:" in the middle of tasks it fails with "does not exist, or is not readable".

**Comment by athak** [comment]
Seeing similar behavior using Ansible 2.3.1.0. The role also has a task file that includes others. Calling the role via `include_role` works fine but using `include_role` with `tasks_from` fails with the reported ` does not exist, or is not readable` error.

**Comment by emestee** [comment]
@athak Same with vars_from in 2.3.1.0

**Comment by dmOx** [comment]
Works in 2.2.1.0 version on Debian, but after upgrade to 2.4.1.0 it breaks like:
```
TASK [nginx_base : Setup for RedHat family] *************************************************************
fatal: [testhost1.local]: FAILED! => {"failed": true, "reason": "Unable to retrieve file contents\nCould not find or access '/home/dmoxx/x/x/ansible/playbook_dir/setup-RedHat.yml'"}
```

When running like
```
$ pwd
/home/dmox/x/x/ansible/x
$ ansible-playbook -i local-lxc.inv playbook.yml
```

This code
```
$ cat roles/nginx_base/tasks/main.yml | head -22
---
# Variable setup.
- name: Include OS-specific variables
  include_vars: "{{ ansible_os_family }}.yml"

- name: Define nginx_posix_user
  set_fact:
    nginx_posix_user: "{{ __nginx_user }}"
  when: nginx_posix_user is not defined

- name: Define nginx_posix_group
  set_fact:
    nginx_posix_group: "{{ __nginx_group }}"
  when: nginx_posix_group is not defined


# Setup/install tasks.
- name: Setup for RedHat family
  include: "setup-RedHat.yml"
  when: ansible_os_family == 'RedHat'
  static: no

```
Role "nginx_base" executed like
"roles" task in playbook -> "include_role" A -> dependency of A -> nginx_base

**Comment by sivel** [comment]
I am unable to reproduce this with the following configuration:

```
- hosts: localhost
  gather_facts: false
  vars:
    user_list:
      - one
      - two
  pre_tasks:
    - include: general-base.yml
```

general-base.yml
```
---
- name: Include proxy role
  include_role:
      name: proxy_install
```

roles/proxy_install/tasks/main.yml
```
---
- include: install.yml

- name: do stuff
  include: do_stuff.yml
  with_items:
    - "{{ user_list }}"
```

roles/proxy_install/tasks/install.yml
```
- debug:
    msg: install
```

roles/proxy_install/tasks/do_stuff.yml
```
- debug:
    var: item
```

This produces:

```
TASK [proxy_install : debug] *****************************************************************************************
ok: [localhost] => {
    "changed": false,
    "msg": "install"
}

TASK [proxy_install : do stuff] **************************************************************************************
included: /Users/matt/projects/ansibledev/playbooks/21328/roles/proxy_install/tasks/do_stuff.yml for localhost
included: /Users/matt/projects/ansibledev/playbooks/21328/roles/proxy_install/tasks/do_stuff.yml for localhost

TASK [proxy_install : debug] *****************************************************************************************
ok: [localhost] => {
    "changed": false,
    "item": "one"
}

TASK [proxy_install : debug] *****************************************************************************************
ok: [localhost] => {
    "changed": false,
    "item": "two"
}
```

Please provide a *full* reproducer that shows this issue.

needs_info

**Comment by sivel** [comment]
I've gone back and tested the reproducer provided in https://github.com/ansible/ansible/issues/21328#issuecomment-302724596

This issue was fixed in e0b0e61b32904975d954fab35ca7ae02147fb6ea and included in the v2.4.0 release.

If you have further questions please stop by IRC or the mailing list:

   * IRC: #ansible on irc.freenode.net
   * mailing list: https://groups.google.com/forum/#!forum/ansible-project
